### PR TITLE
domains: Added Gaussian integers and rationals

### PR DESCRIPTION
**Pull request by jksuom**
#### References to other Issues or PRs

https://github.com/sympy/sympy/issues/15358#issuecomment-428282577

#### Brief description of what is fixed or changed
An implementation Gaussian domains including gcd is added.

#### Release Notes

* polys
    * added new domains for Gaussian integers and rationals

### CODE REVIEW

**Comment by asmeurer** [comment]
Everything looks OK. It would be good to have lots more tests, though, especially for the code that can produce wrong results from typos.

**Comment by czgdp1807** [comment]
@jksuom Is this PR still having a chance to be merged? One, option can be is to raise an issue to keep track of this PR.

**Comment by jksuom** [comment]
More tests are needed. After those pass, I think that this could be merged.

File: sympy/polys/domains/gaussiandomains.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    base = QQ
+    dtype = GaussianRational
+    zero = dtype(0, 0)
+    one = dtype(1, 0)
+
+    rep = 'QQ_I'
+
+    def __init__(self):  # override Domain.__init__
+        """For constructing QQ_I."""
+
```
Comment (kenluck2001) [inline sympy/polys/domains/gaussiandomains.py]: add a pass keyword after the doc string. The code may fail

Comment (jksuom) [reply sympy/polys/domains/gaussiandomains.py]: It looks like the docstring can replace a pass statement.
```
>>> class C(object):
...     def __init__(self):
...         """Docstring."""
... 
>>> c = C()
>>> c
<__main__.C object at 0x7f375a6c8510>
>>>
```

**Comment by czgdp1807** [comment]
@jksuom Please add more tests to this PR, so that we can merge this. Or, someone familiar with the module can extend the PR.

**Comment by czgdp1807** [comment]
ping @jksuom Any updates.

**Comment by czgdp1807** [comment]
Conflicts need to be resolved for this PR.

**Comment by smichr** [comment]
Only the NotImplemented are not covered.

**Comment by smichr** [comment]
@jksuom , could you please check the XXX in the file -- I had a question there.

File: sympy/polys/domains/gaussiandomains.py [hunk 2]
```diff
@@ -276,7 +274,7 @@ def get_ring(self):
 
     def get_field(self):
         """Returns a field associated with ``self``. """
-        return QQ_I
+        return ZZ_I
```
Comment (jksuom) [inline sympy/polys/domains/gaussiandomains.py]: This should be the *field*, i.e. `QQ_I`.

Comment (smichr) [reply sympy/polys/domains/gaussiandomains.py]: The field of ZZ_I is QQ_I and the field of QQ_I is ZZ_I (as you had it before)?

Comment (jksuom) [reply sympy/polys/domains/gaussiandomains.py]: `ZZ_I` is the *ring* of `QQ_I`. It should be returned by `QQ_I.get_ring()`.

File: sympy/polys/domains/gaussiandomains.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    def __init__(self):  # override Domain.__init__
+        """For constructing ZZ_I."""
+
+    def get_ring(self):
+        """Returns a ring associated with ``self``. """
+        return self
+
+    def get_field(self):
+        """Returns a field associated with ``self``. """
+        return ZZ_I
```
Comment (smichr) [inline sympy/polys/domains/gaussiandomains.py]:
```suggestion
        return QQ_I
```

File: sympy/polys/domains/gaussiandomains.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    one = dtype(1, 0)
+
+    rep = 'QQ_I'
+
+    def __init__(self):  # override Domain.__init__
+        """For constructing QQ_I."""
+
+    def get_ring(self):
+        """Returns a ring associated with ``self``. """
+        return QQ_I
```
Comment (smichr) [inline sympy/polys/domains/gaussiandomains.py]:
```suggestion
        return ZZ_I
```

File: sympy/polys/domains/tests/test_domains.py [hunk 5]
```diff
@@ -782,6 +784,60 @@ def test_CC_double():
...
+    assert ZZ_I(3, 0) != 3           # and should this go to Integer?
+    assert QQ_I(S(3)/4, 0) != S(3)/4 # and should this go to Integer?
+    assert ZZ_I(0, 0).quadrant() == 0
+    assert ZZ_I(-1, 0).quadrant() == 2
+    for G in (QQ_I, ZZ_I):
+        q = G(3, 4)
+        assert q._get_xy(pi) == (None, None)
+        assert q._get_xy(2) == (2, 0)
+        assert q._get_xy(2*I) == (0, 2)
+        assert G.get_ring() == G
```
Comment (smichr) [inline sympy/polys/domains/tests/test_domains.py]:
```suggestion
```

File: sympy/polys/domains/tests/test_domains.py [hunk 6]
```diff
@@ -782,6 +784,60 @@ def test_CC_double():
...
+        assert G.get_field() == G
```
Comment (smichr) [inline sympy/polys/domains/tests/test_domains.py]:
```suggestion
```

File: sympy/polys/domains/tests/test_domains.py [hunk 7]
```diff
@@ -782,6 +784,60 @@ def test_CC_double():
...
+        raises(ZeroDivisionError, lambda: divmod(q, 0))
+        raises(ZeroDivisionError, lambda: divmod(q, 0))
+        raises(CoercionFailed, lambda: G.from_sympy(pi))
+        if G == ZZ_I:
+            assert q//3 == G(1, 1)
+            assert 12//q == G(1, -2)
+            assert 12 % q == G(1, 2)
+            assert q % 2 == G(-1, 0)
+            assert i == G(0, 0)
+            assert r == G(2, 0)
```
Comment (smichr) [inline sympy/polys/domains/tests/test_domains.py]:
```suggestion
            assert r == G(2, 0)
            assert G.get_ring() == QQ_I
            assert G.get_field() == G
```

File: sympy/polys/domains/tests/test_domains.py [hunk 8]
```diff
@@ -782,6 +784,60 @@ def test_CC_double():
...
+        else:
+            assert q//3 == G(1, S(4)/3)
```
Comment (smichr) [inline sympy/polys/domains/tests/test_domains.py]:
```suggestion
            assert G.get_ring() == G
            assert G.get_field() == ZZ_I
            assert q//3 == G(1, S(4)/3)
```

**Comment by smichr** [comment]
In summary, here are the minor edits that I made to the orginal PR
```diff
diff --git a/sympy/polys/domains/gaussiandomains.py b/sympy/polys/domains/gaussiandomains.py
index 95d5df2..ee1bc87 100644
--- a/sympy/polys/domains/gaussiandomains.py
+++ b/sympy/polys/domains/gaussiandomains.py
@@ -1,7 +1,5 @@
 """Domains of Gaussian type."""
 
-from __future__ import print_function, division
-
 from sympy.core.basic import Basic
 from sympy.core.numbers import I
 from sympy.polys.polyerrors import CoercionFailed
@@ -118,7 +116,7 @@ def __rtruediv__(self, other):
 
     def __divmod__(self, other):
         if not other:
-            raise ZeroDivisionError('divmod(%s, 0)' % self)
+            raise ZeroDivisionError('divmod({}, 0)'.format(self))
         x, y = self._get_xy(other)
         if x is None:
             return NotImplemented
@@ -153,7 +151,7 @@ def __floordiv__(self, other):
         return qr if qr is NotImplemented else qr[0]
 
     def __rfloordiv__(self, other):
-        qr = self._rdivmod__(self, other)
+        qr = self.__rdivmod__(other)
         return qr if qr is NotImplemented else qr[0]
 
     def __mod__(self, other):
@@ -161,7 +159,7 @@ def __mod__(self, other):
         return qr if qr is NotImplemented else qr[1]
 
     def __rmod__(self, other):
-        qr = self._rdivmod__(self, other)
+        qr = self.__rdivmod__(other)
         return qr if qr is NotImplemented else qr[1]
 
 
@@ -170,7 +168,7 @@ class GaussianRational(GaussianElement):
 
     def __truediv__(self, other):
         if not other:
-            raise ZeroDivisionError('%s / 0' % self)
+            raise ZeroDivisionError('{} / 0'.format(self))
         x, y = self._get_xy(other)
         if x is None:
             return NotImplemented
@@ -184,7 +182,7 @@ def __truediv__(self, other):
     def __rtruediv__(self, other):
         try:
             other = self._parent.convert(other)
-        except CoercienFailed:
+        except CoercionFailed:
             return NotImplemented
         else:
             return other.__truediv__(self)
@@ -197,26 +195,26 @@ def __mod__(self, other):
         except CoercionFailed:
             return NotImplemented
         if not other:
-            raise ZeroDivisionError('%s % 0' % self)
+            raise ZeroDivisionError('{} % 0'.format(self))
         else:
-            return self._parent.zero
+            return self._parent.zero  # XXX always 0?
 
-    def _rmod__(self, other):
+    def __rmod__(self, other):
         try:
             other = self._parent.convert(other)
-        except CoercienFailed:
+        except CoercionFailed:
             return NotImplemented
         else:
-            other.__mod__(self)
+            return other.__mod__(self)
 
     def __divmod__(self, other):
         return self.__truediv__(other), self.__mod__(other)
 
-    def _rdivmod__(self, other):
-        return self._rtruediv__(other), self.__rmod__(other)
+    def __rdivmod__(self, other):
+        return self.__rtruediv__(other), self.__rmod__(other)
 
 
-class GaussianDomain(object):
+class GaussianDomain():
     """Base class for Gaussian domains."""
     base = None  # base domain, ZZ or QQ
 
@@ -239,7 +237,7 @@ def from_sympy(self, a):
         if b is I:
             return self.new(x, y)
         else:
-            raise CoercionFailed("%s is not Gaussian" % a)
+            raise CoercionFailed("{} is not Gaussian".format(a))
 
     def convert(self, element):
         """Convert ``element`` to ``self.dtype``.
```
I then added tests to bring coverage to 92% and only the NotImplemented lines are not covered explicitly.

**Comment by smichr** [comment]
@oscarbenjamin , since you work with number types a bit, do you think you could sign off on this?

**Comment by oscarbenjamin** [comment]
I'm still trying to get my head round how all of this stuff works.

Should it not be possible to do something like:
```julia
In [3]: Poly(I+x, x, domain=QQ_I)                                                                                                 
---------------------------------------------------------------------------
NameError 

In [4]: Poly(I+x, x, domain='QQ_I')                                                                                               
---------------------------------------------------------------------------
OptionError
```
It seems I can if I import the domain explicitly but then
```julia
In [12]: from sympy.polys.domains.gaussiandomains import QQ_I, ZZ_I                                                               

In [13]: p = Poly(I+x, x, domain=QQ_I)                                                                                            

In [14]: p                                                                                                                        
Out[14]: Poly(x + I, x, domain='QQ_I')

In [15]: p * p                                                                                                                    
---------------------------------------------------------------------------
TypeError 
...
TypeError: unsupported operand type(s) for ** or pow(): 'GaussianRational' and 'int'
```

**Comment by jksuom** [comment]
> how all of this stuff works.

To make this work automatically with `I`, it is necessary to add some code to `construct_domain()`.

**Comment by smichr** [comment]
> add some code to `construct_domain()`.

Like "if all the coefficients are integers/rationals, perhaps multiplied by I then use the ZZ_I/QQ_I domain"?

**Comment by oscarbenjamin** [comment]
> Like "if all the coefficients are integers/rationals, perhaps multiplied by I then use the ZZ_I/QQ_I domain"?

Yes. Right now we get
```
In [1]: Poly(x+I)                                                                                                                 
Out[1]: Poly(x + I, x, I, domain='ZZ')

In [2]: Poly(x+I, x)                                                                                                              
Out[2]: Poly(x + I, x, domain='EX')
```
In the first example `I` is treated as a generator. In the second the domain is 'EX' because `construct_domain` hasn't found a domain that involved `I`.

It is possible to construct an extension:
```
In [3]: Poly(x+I, x, extension=True)                                                                                              
Out[3]: Poly(x + I, x, domain='QQ<I>')
```
That gives a an extension field but it should be possible to have the ring `ZZ_I` as domain.

Other domains are imported at top-level in sympy as ZZ etc but can also be passed as strings:
```
In [5]: Poly(x+1, x, domain='QQ')                                                                                                 
Out[5]: Poly(x + 1, x, domain='QQ')
```
So I guess somewhere there is code to recognise the strings.

Having the domain automatically detected in `construct_domain` is what will mean that it actually gets used. I think that will then lead to test failures etc and show which parts of the domain need fixing.

**Comment by jksuom** [comment]
> if all the coefficients are integers/rationals, perhaps multiplied by I then use the ZZ_I/QQ_I domain

I think this how it should be. This should come before algebraics are tested so also ZZ_I could be found if the coefficients are integers. Something like this:

    for coeff in coeffs:
        is_complex = pure_complex(coeff)
        if is_complex:
            # set rationals if either part is rational
            # nothing if both are integers, and
            gaussians = True
        elif coeff.is_Rational:
            # continue possibly resetting gaussians to False

At the end of `_construct_simple` something like this:
                   
        else:
            if gaussians:
                if opt.field or rationals:
                    domain = QQ_I
                else:
                    domain = ZZ_I
            elif opt.field or rationals:
                domain = QQ
            else:
                domain = ZZ

**Comment by czgdp1807** [comment]
@smichr @jksuom @oscarbenjamin Is this ready to go?

**Comment by jksuom** [comment]
There is no support for polynomials yet but it is possible to compute with the domain elements. It think that this could be merged. Polynomial support can be added later.

**Comment by oscarbenjamin** [comment]
I might try to incorporate the domain as poly and see if I can push that here. I suspect that that would give a more robust test of the implementation here.

**Comment by oscarbenjamin** [comment]
Actually we can just merge this for now and I'll do that after.

**Comment by jksuom** [comment]
Thanks!

**Pull request merged**